RealmSwift/ThreadSafeReference.swift

```diff
@@ -122,6 +122,76 @@ public protocol ThreadConfined {
     }
 }
 
+// MARK: ThreadSafe propertyWrapper
+
+/**
+    A property wrapper type that may be passed between threads.
+
+    A `@ThreadSafe` property contains a thread-safe reference to the underlying wrapped value.
+    This reference is resolved to the thread on which the wrapped value is accessed. A new thread
+    safe reference is created each time the property is accessed.
+
+ - warning: This property wrapper should not be used for properties on long lived objects.
+            `@ThreadSafe` properties contain a `ThreadSafeReference` which
+            can pin the source version of the Realm in use. This means that this property
+            wrapper is **better suited for function arguments and local variables**
+            **that get captured by an aynchronously dispatched block.**
+
+ - see: `ThreadSafeReference`
+ - see: `ThreadConfined`
+*/
+@propertyWrapper public class ThreadSafe<T: ThreadConfined> {
+    private var threadSafeReference: ThreadSafeReference<T>?
+    private var rlmConfiguration: RLMRealmConfiguration?
+    private var barrier = DispatchQueue(label: "thread-safe-property-wrapper")
```

Comment:
Creating a queue for every property wrapper is a *lot* of overhead. GCD queues were supposed to be cheap to create, but that didn't work out in practice. https://gist.github.com/tclementdev/6af616354912b0347cdf6db159c37057 suggests that this should either be a static queue shared between all of them or just a NSLock.